src/cargo/ops/resolve.rs

```diff
@@ -308,12 +308,13 @@ pub fn resolve_with_previous<'gctx>(
     }
 
     let avoid_patch_ids = if register_patches {
-        register_patch_entries(registry, ws, previous, &mut version_prefs, pre_patch_keep)?
+        register_patch_entries(registry, ws, previous, &mut version_prefs, keep)?
     } else {
         HashSet::new()
     };
 
-    let keep = |p: &PackageId| pre_patch_keep(p) && !avoid_patch_ids.contains(p);
+    // Refine `keep` with patches that should avoid locking.
+    let keep = |p: &PackageId| keep(p) && !avoid_patch_ids.contains(p);
```

Comment:
While I'm normally a fan of shadowing, I feel like decorating a closure with another closure is likely going to make this easy to miss which could make reading the code confusing